marshal/marshal.go

```diff
@@ -0,0 +1,212 @@
+package marshal
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"golang.org/x/exp/maps"
+)
+
+const force_send_field_name = "ForceSendFields"
+
+// Marshal returns a JSON encoding of the given object. Included fields:
+// - non-empty value
+// - a basic type whose field's name is present in forceSendFields
+// Embedded structs are still considered a separate struct. ForceSendFields
+// in an embedded struct only impact the fields of the embedded struct.
+// Conversely, an embedded struct is not impacted by the ForceSendFields
+// of the struct containing it.
+func Marshal(object any) ([]byte, error) {
+
+	dataMap, err := structAsMap(object)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(dataMap)
+}
+
+// Converts the object to a map
+func structAsMap(object any) (map[string]any, error) {
+	result := make(map[string]any)
+
+	// If the object is nil or a pointer to nil, don't do anything
+	if object == nil || (reflect.ValueOf(object).Kind() == reflect.Ptr && reflect.ValueOf(object).IsNil()) {
+		return result, nil
+	}
+
+	value := reflect.ValueOf(object)
+	value = reflect.Indirect(value)
+	objectType := value.Type()
+
+	includeFields, err := getForceSendFields(object, objectType.Name())
+
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < value.NumField(); i++ {
+		jsonTag := objectType.Field(i).Tag.Get("json")
+		tag, err := parseJSONTag(jsonTag)
+		if err != nil {
+			return nil, err
+		}
+
+		fieldValue := value.Field(i)
+		fieldType := objectType.Field(i)
```

Comment:
Doing this every time is pretty expensive. Please think about caching a bunch of this information per type (e.g. the list of field names, the tags, etc are all static per type).

Reply:
I added some caching and did a performance test. 
Dataset: Repeat single test 1.000.000 times 
No cache: ~10.5s
Cache: ~8s

Reply:
Does this also include the stuff to get the index of the force send fields field, etc? The comment was not about this line alone, about caching type info in general.